// league-of-legends-api/endpoints/summoner.js
import consts from '../consts.js';

/**
 * This class implements summoner endpoints of the LoL API <https://developer.riotgames.com/api-methods/#summoner-v3>
 * @author Filip Butkiewicz aka. bagnz0r <http://github.com/bagnz0r>
 */
class Summoner {

    constructor(apiKey, region) {
        this.headers = new Headers({
            'X-Riot-Token': apiKey
        });

        this.url = `https://${consts.zone[region]}/lol/summoner/v3`;
    }

    /**
     * Get a summoner by account ID.
     * @param {Number} accountId
     * @return {Promise}
     */
    getSummonerByAccount(accountId) {
        return fetch(`${this.url}/summoners/by-account/${accountId}`, { headers: this.headers });
    }

    /**
     * Get a summoner by summoner name.
     * @param {String} summonerName
     * @return {Promise}
     */
    getSummonerByName(summonerName) {
        return fetch(`${this.url}/summoners/by-name/${summonerName}`, { headers: this.headers });
    }

    /**
     * Get a summoner by summoner ID.
     * @param {Number} summonerById
     * @return {Promise}
     */
    getSummonerById(summonerById) {
        return fetch(`${this.url}/summoners/${summonerId}`, { headers: this.headers });
    }

}

export default Summoner;
