// league-of-legends-api/endpoints/champion-mastery.js
import consts from '../consts.js';

/**
 * This class implements champion-mastery endpoints of the LoL API <https://developer.riotgames.com/api-methods/#champion-mastery-v3>
 * @author Filip Butkiewicz aka. bagnz0r <http://github.com/bagnz0r>
 */
class ChampionMastery {

    constructor(apiKey, region, summonerId) {
        this.headers = new Headers({
            'X-Riot-Token': apiKey
        });

        this.summonerId = summonerId;
        this.url = `https://${consts.zone[region]}/lol/champion-mastery/v3`;
    }

    /**
     * Get all champion mastery entries sorted by number of champion points descending.
     * @return {Promise}
     */
    getMasteries() {
        return fetch(`${this.url}/champion-masteries/by-summoner/${this.summonerId}`, { headers: this.headers });
    }

    /**
     * Get a champion mastery by player ID and champion ID
     * @param {Number} championId
     * @return {Promise}
     */
    getMasteriesByChampion(championId) {
        return fetch(`${this.url}/champion-masteries/by-summoner/${this.summonerId}/by-champion/${championId}`, { headers: this.headers });
    }

    /**
     * Get a player's total champion mastery score, which is the sum of individual champion mastery levels.
     * @return {Promise}
     */
    getScores() {
        return fetch(`${this.url}/scores/by-summoner/${this.summonerId}`, { headers: this.headers });
    }

}

export default ChampionMastery;
